File: projects/06/assembler/src/parser.py
"""
A parser for the assembler.

Responsible for parsing an input assembly language program and providing
convenient access to the commands used in the program.
"""

from . import utils
from typing import TextIO


class Parser:
  """
  A parser for the assembler.

  Encapsulates access to the input code. Reads an assembly propgram and
  provides access to its components.

  Attributes:
    filename (str): The filename/path of the assembly language file.
    file (TextIO): The text file that has been opened for reading.
    currentCommand (str): The current command of the assembly file.
  """

  def __init__(self, file: TextIO) -> None:
    """
    Initializes the parser based on an assembly file.

    Args:
      file (TextIO): The assembly file opened in read mode.
    """
    self._file = file
    self.currentCommand = None

  def hasMoreCommands(self) -> bool:
    """
    Checks if there are more commands in the input.

    Returns:
      A bool representing whether or not there are more commands in the input
      file from the current line. Note that whitespace and comments do not
      count.
    """
    return utils.peek(self._file)

  def advance(self) -> None:
    """
    Reads the next command from the file and makes it the current command. Only
    works when `hasMoreCommands` is true.
    """
    if self.hasMoreCommands():
      self.currentCommand = self._file.readline()
    else:
      self.currentCommand = None

  def commandType(self) -> utils.COMMAND_TYPE:
    """
    Checks the type of command that the current command is set to.

    Returns a `COMMAND_TYPE` enumeration depending on which command is set.
    """
    if not self.currentCommand:
      return utils.COMMAND_TYPE.NON_COMMAND
    elif self.currentCommand.strip() == "":
      return utils.COMMAND_TYPE.NON_COMMAND
    elif self.currentCommand.strip()[0] == '@':
      return utils.COMMAND_TYPE.A_COMMAND
    elif self.currentCommand[0] == '(':
      return utils.COMMAND_TYPE.L_COMMAND
    elif self.currentCommand.strip()[0] == '/':
      return utils.COMMAND_TYPE.NON_COMMAND
    return utils.COMMAND_TYPE.C_COMMAND

  def symbol(self) -> str:
    """
    Returns the symbol of the current command.

    Returns a string representing the symbol of the current command. Only works
    for A or L commands.
    """
    if not self.currentCommand:
      return
    commandType = self.commandType()
    line = self.currentCommand.strip()
    if commandType == utils.COMMAND_TYPE.A_COMMAND:
      return line[1:].strip()
    elif commandType == utils.COMMAND_TYPE.L_COMMAND:
      return line[1:-1]

  def dest(self) -> str:
    """
    Returns the destination of the current command.

    Returns a string representing the `dest` mnemonic. Only works for C
    commands.
    """
    i = self.currentCommand.find('=')
    if self.commandType() == utils.COMMAND_TYPE.C_COMMAND and i > 0:
      return self.currentCommand[:i].strip()
    return ""

  def comp(self) -> str:
    """
    Returns the computation of the current command.

    Returns a string representing the `comp` mnemonic. Only works for C
    commands.
    """
    if self.commandType() == utils.COMMAND_TYPE.C_COMMAND:
      i = self.currentCommand.find('=') + 1
      j = self.currentCommand.find(';')
      return self.currentCommand[i:j].strip()

  def jump(self) -> str:
    """
    Returns the jump of the current command.

    Returns a string representing the `jump` mnemonic. Only works for C
    commands.
    """
    if self.commandType() == utils.COMMAND_TYPE.C_COMMAND:
      i = self.currentCommand.find(';')
      if i != -1:
        return self.currentCommand[i + 1:].strip()
      return ""

  def reset(self) -> None:
    """
    Resets the file pointer to the beginning.
    """
    self._file.seek(0)

File: projects/06/assembler/parser.cpp
/**
 * Created by Kush Padalia on May 24, 2024
 */

#include "parser.h"
#include <iostream>
#include "commandType.h"
#include <boost/algorithm/string.hpp>

namespace assembler {
  parser::parser(const std::string &fileName) {
    file.open(fileName);
  }

  parser::~parser() {
    file.close();
  }

  void parser::advance() {
    (void) std::getline(file, currentCommand);
  }

  const std::string &parser::getCurrentCommand() const {
    return currentCommand;
  }

  assembler::commandType parser::commandType() const {
    const std::string trimmedCurrentCommand = boost::trim_copy(currentCommand);
    assembler::commandType commandType;
    if (const int i = 0; currentCommand.empty() || trimmedCurrentCommand.empty() || trimmedCurrentCommand[i] == '/') {
      commandType = assembler::commandType::NonCommand;
    } else if (trimmedCurrentCommand[i] == '@') {
      commandType = assembler::commandType::ACommand;
    } else if (trimmedCurrentCommand[i] == '(') {
      commandType = assembler::commandType::LCommand;
    } else {
      commandType = assembler::commandType::CCommand;
    }
    return commandType;
  }

  std::string parser::symbol() const {
    std::string symbol;
    assembler::commandType commandType = parser::commandType();
    std::string command = boost::trim_copy(currentCommand);
    if (commandType == assembler::commandType::ACommand) {
      symbol = boost::trim_copy(command.substr(1));
    } else if (commandType == assembler::commandType::LCommand) {
      symbol = command.substr(1, command.size() - 2);
    } else {
      symbol = "";
    }
    return symbol;
  }

  std::string parser::dest() const {
    std::basic_string<char>::size_type equalPos = currentCommand.find('=');
    std::string dest;
    if (parser::commandType() == assembler::commandType::CCommand && equalPos < currentCommand.size()) {
      dest = boost::trim_copy(currentCommand.substr(0, equalPos));
    } else {
      dest = "";
    }
    return dest;
  }

  std::string parser::comp() const {
    std::string comp;
    if (parser::commandType() == assembler::commandType::CCommand) {
      const size_t i = currentCommand.find('=') + static_cast<size_t>(1);
      const size_t j = currentCommand.find(';');
      comp = boost::trim_copy(currentCommand.substr(i, (j - i)));
    } else {
      comp = "";
    }
    return comp;
  }

  std::string parser::jump() const {
    std::string jump;
    if (parser::commandType() == assembler::commandType::CCommand) {
      const std::basic_string<char>::size_type i = currentCommand.find(';');
      jump = (i != -1) ? (boost::trim_copy(currentCommand.substr(i + 1))) :
             "";
    } else {
      jump = "";
    }
    return jump;
  }

  void parser::reset() {
    file.clear();
    (void) file.seekg(0);
  }

  bool parser::hasMoreCommands() {
    return file.peek() != EOF;
  }
}
